### PR TITLE
Add option to output .js files while preserving jsx

### PR DESCRIPTION
**Pull request by ds300**
This PR adds the ability to preserve jsx in source code, but also to output `.js` files rather than `.jsx` files. This is useful for react-native which does not support `.jsx` files.

This is a replacement PR for #9864 which seems to have stalled. It also solves #11158.

### CODE REVIEW

**Comment by RyanCavanaugh** [comment]
Thanks! We were thinking the option name should just be `react-native` for succinctness/clarity.

**Comment by ds300** [comment]
A much better name :) I'll make the change later tonight.

**Comment by sheetalkamat** [comment]
You also need to enable notification for es5 transformation for new option. Refer: https://github.com/Microsoft/TypeScript/blob/master/src/compiler/transformers/es5.ts#L17

**Comment by sheetalkamat** [comment]
Also please add tests in tests\cases\compiler for easy baseline verification of this scenario. Thank you

**Comment by ds300** [comment]
@sheetalkamat I think I made the appropriate change in `es5.ts`, but am not sure whether I'm doing the test cases correctly. There doesn't seem to be any notifications in the baseline outputs anyway. Do you mind taking a look?

File: tests/cases/fourslash/getEmitOutputTsxFile_ReactNative.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
+/// <reference path="fourslash.ts" />
```
Comment (RyanCavanaugh) [inline tests/cases/fourslash/getEmitOutputTsxFile_ReactNative.ts]: This shouldn't be a fourslash test (they are much slower than compiler baselines). See other tests in `tests/cases/compiler` for how to write a multi-file baseline test (search "`@filename`")

Comment (ds300) [inline tests/cases/fourslash/getEmitOutputTsxFile_ReactNative.ts]: The other compiler testcases I made after this also test the PR's behaviour, so I suppose I can just delete this fourslash one.

File: lib/protocol.d.ts [hunk 2]
```diff
@@ -1798,9 +1798,10 @@ declare namespace ts.server.protocol {
     namespace JsxEmit {
         type None = "None";
         type Preserve = "Preserve";
+        type ReactNative = "ReactNative";
```
Comment (mhegazy) [inline lib/protocol.d.ts]: can you revert the changes to this file, the file is generated as part of the build. you have edited the correct file under src/server/protocol.ts
